src/generic/grid.cpp

```diff
@@ -11337,4 +11400,617 @@ wxGetContentRect(wxSize contentSize,
     return contentRect;
 }
 
+
+#if wxUSE_ACCESSIBILITY
+
+//-----------------------------------------------------------------------------
+// helpers for wxGridAccessible and wxGridCellAccessible
+//-----------------------------------------------------------------------------
+
+namespace
+{
+
+// Helper for converting between grid coordinates and child IDs.
+//
+// Child IDs are 1-based and are used to identify individual cells in the grid
+// in the top-down, left-to-right order.
+
+// Return the child ID corresponding to the given grid cell coordinates.
+int ConvertCoordsToChildId(wxGrid* grid, wxGridCellCoords coords)
+{
+    int childId = 0;
+    int numberCols = grid->GetNumberCols();
+    if ( grid->GetRowLabelSize() )
+        numberCols++;
+
+    if ( !coords )
+        childId = 1;
+    else if ( coords.GetRow() == -1 )
+    {
+        childId = coords.GetCol() + 1;
+        if ( grid->GetRowLabelSize() )
+            childId++;
+    }
+    else if ( coords.GetCol() == -1 )
+    {
+        childId = coords.GetRow() * numberCols + 1;
+        if ( grid->GetColLabelSize() )
+            childId += numberCols;
+    }
+    else
+    {
+        childId = coords.GetRow() * numberCols + coords.GetCol() + 1;
+        if ( grid->GetColLabelSize() )
+            childId += numberCols;
+        if ( grid->GetRowLabelSize() )
+            childId++;
+    }
+    return childId;
+}
+
+// Get coordinates corresponding to the given child ID (see the converse
+// function above).
+wxGridCellCoords ConvertChildIdToCoords(wxGrid* grid, int childId)
+{
+    wxGridCellCoords coords;
+
+    if ( childId != 1 || !grid->GetColLabelSize() || !grid->GetRowLabelSize() )
+    {
+        int numCols = grid->GetNumberCols();
+        if ( grid->GetRowLabelSize() )
+            numCols++;
+        int row = (childId - 1) / numCols;
+        int col = (childId - 1) % numCols;
```

Comment:
Are we sure that `numCols` can't be 0 here? It doesn't seem totally obvious...

Reply:
GetChildCount should return the correct value in this case, i.e. 0 if there is also no header.
But yes, I don't know the timing and whether the accessibility framework's data could become out of date.

Reply:
Just to be clear: what I meant was that we must ensure we don't divide by 0 here as this would crash the program. If we're not sure that this code can't be called in the case when it's 0, we must check for it here explicitly.

Reply:
Yes, that's how I understood your comment. Is there a good way to return a NULL value instead of the wxGridCellCoords instance?

When thinking about this, it might not be the worst idea to validate the coordinates at some other positions, e.g. at `wxGridCellAccessible::GetName`. I'm not sure how the accessibility framework is actually working. If it retrieves the information and caches it, things should be fine. If a client application like NVDA could trigger a late call itself, then a check would be good.

I think I will implement this.

Reply:
No, we can't return null object. We could change the function to return `unique_ptr<wxGridCellCoords>` which could be null, but this looks like an overkill — it looks like we could just return invalid `wxGridCellCoords` (i.e. `-1,-1`) for a completely empty grid, couldn't we?

But we really, really shouldn't crash with a division by 0.

Reply:
The cleanest approach for all would be to destroy the current wxGridCellAccessible instance on each change of table dimension or label sizes, but I don't see a mechanism in wxGrid for this, especially in the general case with a separate data table.

Returning `(-1,-1)` for `numCols==0` is probably good enough.

The correct but clumsy approach would be for `wxGridCellAccessible` to cache the number of rows, cols and sizes of row and col header at creation. This could then be compared to the current dimensions and in case of mismatch error values would be returned. This seems a bit overengineered, though. I've modeled this after the implemenation in DataViewCtrl and I don't remember any checks there.

Reply:
Should I push this modification?
```C++
// Get coordinates corresponding to the given child ID (see the converse
// function above).
wxGridCellCoords ConvertChildIdToCoords(wxGrid* grid, int childId)
{
    wxGridCellCoords coords;

    if ( childId != 1 || !grid->GetColLabelSize() || !grid->GetRowLabelSize() )
    {
        int numCols = grid->GetNumberCols();
        if ( grid->GetRowLabelSize() )
            numCols++;
        if ( numCols ) {
            // actually, numCols == 0 should not happen
            int row = (childId - 1) / numCols;
            int col = (childId - 1) % numCols;
            if ( grid->GetRowLabelSize() )
            {
                if ( col == 0 )
                    col = -1;
                else
                    col--;
            }
            if ( grid->GetColLabelSize() )
            {
                if ( row == 0 )
                    row = -1;
                else
                    row--;
            }
            coords.SetCol(col);
            coords.SetRow(row);
        }
    }

    return coords;
}
```

Reply:
If it really shouldn't happen, I'd just add `wxCHECK_MSG( numCols, coords, "shouldn't be called" );` to make it clear. But if it can happen, i.e. if this function can be called because the user uses screen reader on an empty grid, then I'd indeed do this but change the comment.

Reply:
An empty grid is fine.
`HitTest` and `GetChildCount` are responsible for returning only valid child ids. In case of a grid with zero columns, the child count would be the number of rows plus 1 for the corner.
The only way to have numCols==0 here would be if a call would be done on an outdated child id due to a grid resize. I don't think that the accessibility mechanism does that, but I would not bet my life on it. So, the `wxCHECK_MSG` seems a good containment for the hypothetical case.

Btw. we would be in good company.
I found that Visual Studio crashes a lot when using a screen reader: 
https://developercommunity.visualstudio.com/t/Many-crashes-when-accessibility-tools-ar/10606336